spec.bs

```diff
@@ -8349,6 +8665,104 @@ scoring signals requests=] |entriesToBatch|:
+
+1. Let |firstRequest| be |entriesToBatch|[0].
+1. Let |compressionGroups| be an empty [=list=], whose [=list/items=] are [=scoring compression groups=].
+1. Let |compressionGroupMap| be an empty [=map=], where the [=map/keys=] are integers as
+  [=scoring compression group/compression group id=], and the [=map/values=] are [=maps=] with integers as [=map/keys=]
+  for [=scoring partition/id=], and [=scoring partitions=] as their [=map/values=].
+1. Let |compressionIdMap| be an empty [=map=], whose [=map/keys=] are [=tuples=] of ([=origin=],
+  [=origin=]) and [=map/values=] are integers.
+1. Let |renderUrlIdMap| be an empty [=map=], whose [=map/keys=] are [=URLs=] and [=map/values=] are [=tuples=]
+  of (interger, integer).
+1. Let |nextCompressionGroupId| be 0.
+1. [=map/For each=] |request| of |entriesToBatch|:
+  1. Let |mapKey| be a tuple of (|request|'s [=trusted scoring signals request/owner origin=],
+    |request|'s [=trusted scoring signals request/joining origin=]).
+  1. If |compressionIdMap| does not [=map/contain=] |mapKey|:
+    1. Set |compressionIdMap|[|mapKey|] to |nextCompressionGroupId|.
+    1. Increment |nextCompressionGroupId| by 1.
+  1. Let |compressionGroupId| be |compressionIdMap|[|mapKey|].
+  1. If |compressionGroupMap| does not [=map/contain=] |compressionGroupId|:
+    1. Set |compressionGroupMap|[|compressionGroupId|] to an empty [=map=], whose [=map/keys=] are integers
+  1. Let |partitionId| be [=list/size=] of |compressionGroupMap|[|compressionGroupId|].
+  1. Set |renderUrlIdMap|[|request|'s [=trusted scoring signals request/render URL=]] to [=tuple=]
+    of (|compressionGroupId|, |partitionId|).
+  1. Set |compressionGroupMap|[|compressionGroupId|][|partitionId|] with the following [=struct/items=]:
+    : [=scoring partition=]
+    :: a [=struct=] with the following [=struct/items=]:
+      : [=scoring partition/id=]
+      :: an integer records the identifier of this [=scoring partition=].
+
+      : [=scoring partition/namespace=]
+      :: a [=map=] whose [=map/keys=] are [=strings=] and [=map/values=] are [=list=] of [=strings=].
+
+      : [=scoring partition/metadata=]
+      :: a [=map=] whose [=map/keys=] and [=map/values=] are [=strings=].
+
+    1. Set |compressionGroupMap|[|compressionGroupId|][|partitionId|]'s [=scoring partition/namespace=]["render_urls"]
+      to [|request|'s [=trusted scoring signals request/render URL=]].
+    1. Set |compressionGroupMap|[|compressionGroupId|][|partitionId|]'s [=scoring partition/namespace=]["ad_component_render_urls"]
+      to |request|'s [=trusted scoring signals request/ad component URLs=].
+    1. Set |compressionGroupMap|[|compressionGroupId|][|partitionId|]'s [=scoring partition/metadata=]["experiment_group_id"]
+      to |firstRequest|'s [=trusted scoring signals request/seller experiment group id=].
+1. [=map/For each=] |id| → |group| in |compressionGroupMap|:
+  1. Let |compressionGroup| be an empty [=scoring compression group=].

```

Comment:
a struct has no concept of "empty". You either need to set a field's value, or have a default value for a field in the struct's definition.
You can simplify this as "let |...| be a [=...=] whose [=.../compression group id=] is |id|, and [=...partition=] is an empty [=list=]".